File: src/qq/qq-cli/commands/hunt/facf/distribution.rs
// SPDX-FileCopyrightText: © 2024-2026 Austin Seipp
// SPDX-License-Identifier: Apache-2.0

/// The probability distribution over suspects + "no culprit".
///
/// For `n` suspects, the distribution has length `n+1`:
/// - `probs[0..n]` = probability each suspect is the culprit
/// - `probs[n]` = probability there is no culprit (original failure was a flake)
#[derive(Debug, Clone)]
pub struct Distribution {
    probs: Vec<f64>,
}

impl Distribution {
    /// Create a uniform prior distribution for `n` suspects.
    ///
    /// Each suspect (and "no culprit") gets probability `1/(n+1)`.
    pub fn uniform(num_suspects: usize) -> Self {
        let prob = 1.0 / (num_suspects + 1) as f64;
        Self {
            probs: vec![prob; num_suspects + 1],
        }
    }

    /// Create a distribution from raw probabilities.
    ///
    /// The probabilities should sum to 1.0 (within tolerance).
    pub fn from_probs(probs: Vec<f64>) -> Self {
        assert!(probs.len() >= 2, "need at least 1 suspect + no-culprit");
        assert!(
            probs.iter().all(|&p| p >= 0.0),
            "all probabilities must be non-negative"
        );
        Self { probs }
    }

    /// Get probability for suspect `i` (or no-culprit if `i == num_suspects`).
    pub fn get(&self, i: usize) -> f64 {
        self.probs[i]
    }

    /// Get index with maximum probability.
    pub fn argmax(&self) -> usize {
        self.probs
            .iter()
            .enumerate()
            .max_by(|(_, a), (_, b)| a.partial_cmp(b).unwrap())
            .map(|(i, _)| i)
            .unwrap_or(0)
    }

    /// Get maximum probability value.
    pub fn max(&self) -> f64 {
        self.probs.iter().cloned().fold(0.0, f64::max)
    }

    /// Compute cumulative distribution function.
    ///
    /// Returns a vector where `cdf[i]` = sum of `probs[0..=i]`.
    pub fn cdf(&self) -> Vec<f64> {
        let mut cdf = Vec::with_capacity(self.probs.len());
        let mut cumsum = 0.0;
        for &p in &self.probs {
            cumsum += p;
            cdf.push(cumsum);
        }
        cdf
    }

    /// Number of suspects (excludes no-culprit slot).
    pub fn num_suspects(&self) -> usize {
        self.probs.len() - 1
    }

    /// Get the raw probability vector.
    pub fn probs(&self) -> &[f64] {
        &self.probs
    }

    /// Get a mutable reference to the raw probability vector.
    pub fn probs_mut(&mut self) -> &mut [f64] {
        &mut self.probs
    }

    /// Check if probabilities sum to 1.0 (within tolerance).
    pub fn is_normalized(&self) -> bool {
        let sum: f64 = self.probs.iter().sum();
        (sum - 1.0).abs() < 1e-9
    }

    /// Normalize the distribution so probabilities sum to 1.0.
    pub fn normalize(&mut self) {
        let sum: f64 = self.probs.iter().sum();
        if sum > 0.0 {
            for p in &mut self.probs {
                *p /= sum;
            }
        }
    }
}

File: tilde/aseipp/orchestra/src/facf/distribution.ts
// SPDX-FileCopyrightText: © 2024-2026 Austin Seipp
// SPDX-License-Identifier: Apache-2.0

/**
 * Probability-vector operations for the pure FACF search engine.
 *
 * A vector has one slot per suspect, oldest to newest, followed by a final
 * "no culprit" hypothesis. This ports `facf/distribution.rs` from revision
 * `zrqw` (commit `f28b0530f45979e1eed4be25c9a694fef98cbcbf`). Unlike the Rust
 * mutable-slice accessor, this module only exposes copies, so readers cannot
 * silently corrupt the search state. Inputs additionally reject non-finite
 * values. No celld, repository, or execution-platform concepts live here.
 *
 * @module
 */

/** A finite, nonnegative vector over suspects plus the no-culprit hypothesis. */
export class Distribution {
  /** Private vector; callers obtain defensive copies through `probs`. */
  readonly #probs: number[];

  /** Constructs a validated vector; normalization is explicit. */
  private constructor(probs: readonly number[]) {
    if (probs.length < 2) {
      throw new RangeError("need at least 1 suspect + no-culprit");
    }
    if (
      !probs.every((probability) =>
        Number.isFinite(probability) && probability >= 0
      )
    ) {
      throw new RangeError("all probabilities must be finite and non-negative");
    }
    this.#probs = [...probs];
  }

  /** Gives each suspect and no-culprit equal prior probability. */
  static uniform(numSuspects: number): Distribution {
    if (!Number.isSafeInteger(numSuspects) || numSuspects < 1) {
      throw new RangeError("must have at least one suspect (a safe integer)");
    }
    return new Distribution(Array(numSuspects + 1).fill(1 / (numSuspects + 1)));
  }

  /** Copies raw weights, which need not sum to one until `normalize` is called. */
  static fromProbs(probs: readonly number[]): Distribution {
    return new Distribution(probs);
  }

  /** Number of ordered suspect slots, excluding no-culprit. */
  get numSuspects(): number {
    return this.#probs.length - 1;
  }

  /** Returns an independent copy of the complete vector. */
  get probs(): number[] {
    return [...this.#probs];
  }

  /** Reads a suspect probability, or no-culprit at index `numSuspects`. */
  get(index: number): number {
    if (
      !Number.isSafeInteger(index) || index < 0 || index >= this.#probs.length
    ) {
      throw new RangeError("distribution index out of range");
    }
    return this.#probs[index];
  }

  /** Returns the last maximum, matching Rust Iterator::max_by tie-breaking. */
  argmax(): number {
    let winner = 0;
    for (let index = 1; index < this.#probs.length; index++) {
      if (this.#probs[index] >= this.#probs[winner]) winner = index;
    }
    return winner;
  }

  /** Returns the probability of the winning hypothesis. */
  max(): number {
    return this.#probs[this.argmax()];
  }

  /** Returns inclusive cumulative probability, including the no-culprit slot. */
  cdf(): number[] {
    let sum = 0;
    return this.#probs.map((probability) => sum += probability);
  }

  /** Tests normalization using the Rust implementation's absolute tolerance. */
  isNormalized(): boolean {
    return Math.abs(
      this.#probs.reduce((sum, probability) => sum + probability, 0) - 1,
    ) < 1e-9;
  }

  /**
   * Normalizes positive weights in place; all-zero weights remain all zero.
   * Scaling by the largest weight first avoids overflow for large finite inputs.
   */
  normalize(): void {
    const maximum = this.max();
    if (maximum === 0) return;
    const total = this.#probs.reduce(
      (sum, probability) => sum + probability / maximum,
      0,
    );
    for (let index = 0; index < this.#probs.length; index++) {
      this.#probs[index] = (this.#probs[index] / maximum) / total;
    }
  }
}
